services/worker-manager/src/providers/azure.js

```diff
@@ -0,0 +1,379 @@
+
+      let providerData = {
+        location: cfg.location,
+        resourceGroupName: this.providerConfig.resourceGroupName,
+        vm: {
+          name: virtualMachineName,
+          location: cfg.location,
+        },
+        ip: {
+          name: ipAddressName,
+          location: cfg.location,
+        },
+        nic: {
+          name: networkInterfaceName,
+          location: cfg.location,
+        },
+        disk: {
+          name: diskName,
+          location: cfg.location,
+        },
+      };
+
+      try {
+        // create NIC, public IP, and VM
+        ipAddress = await this._enqueue('query', () => this.networkClient.publicIPAddresses.createOrUpdate(resourceGroupName, ipAddressName, {
+          location: cfg.location,
+          publicIPAllocationMethod: 'Dynamic',
+        }));
+
+        networkInterface = await this._enqueue('query', () => this.networkClient.networkInterfaces.createOrUpdate(resourceGroupName, networkInterfaceName, {
+          location: cfg.location,
+          ipConfigurations: [
+            {
+              name: ipAddressName,
+              privateIPAllocationMethod: 'Dynamic',
+              subnet: {
+                id: cfg.subnetId,
+              },
+              publicIPAddress: {
+                id: ipAddress.id,
+              },
+            },
+          ],
+        }));
+
+        const customData = Buffer.from(JSON.stringify({
+          workerPoolId,
+          providerId: this.providerId,
+          workerGroup: this.providerId,
+          rootUrl: this.rootUrl,
+          workerConfig: cfg.workerConfig || {},
+        })).toString('base64');
+
+        virtualMachine = await this._enqueue('query', () => this.computeClient.virtualMachines.createOrUpdate(resourceGroupName, virtualMachineName, {
+          ...cfg,
+          osProfile: {
+            adminUsername: 'worker',
+            ...cfg.osProfile,
+            // throw this away
+            adminPassword: slugid.nice(),
+            computerName: virtualMachineName,
+            customData,
+          },
+          networkProfile: {
+            ...cfg.networkProfile,
+            networkInterfaces: [
+              {
+                id: networkInterface.id,
+                primary: true,
+              },
+            ],
+          },
+          priority: cfg.priority || "Spot",
+          evictionPolicy: cfg.evictionPolicy || "Deallocate",
+          billingProfile: {
+            "maxPrice": -1,
+            ...cfg.billingProfile,
+          },
+          tags: {
+            ...cfg.tags || {},
+            'created-by': `taskcluster-wm-${this.providerId}`.replace(/[^a-zA-Z0-9-]/g, '-'),
+            'managed-by': 'taskcluster',
+            'worker-pool-id': workerPoolId.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase(),
+            'owner': workerPool.owner.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase(),
+          },
+        }));
+      } catch (err) {
+        await this._removeWorker({...providerData});
```

Comment:
Let's add a comment that this is here to clean up the resources like nics, etc and that if they don't exist yet it is fine.